### PR TITLE
Add docstring checker to pre-commit

### PR DESCRIPTION
**Pull request by Cjkjvfnby**
### Describe your change:

- Add flake8 docstring to the pre-commit hook
- Add global configuration for checks
- Add per-file exclusion to the violation.
- Minor doc fixes

* [ ] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [x] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [ ] This pull request is all my own work -- I have not plagiarized.
* [ ] I know that pull requests will not be merged if they fail the automated tests.
* [ ] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [ ] All filenames are in all lowercase characters with no spaces or dashes.
* [ ] All functions and variable names follow Python naming conventions.
* [ ] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [ ] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

The script I was using to generate `per-file-ignore` list.  (copy paste pre-commit output into a)

```python
a = '''
ciphers/vigenere_cipher.py:1:1: D100 Missing docstring in public module
ciphers/vigenere_cipher.py:4:1: D103 Missing docstring in public function
'''.strip()

result = {}

for line in a.split("\n"):
    file_name, _, tail = line.partition(":")
    _, _, warning = tail.rpartition(": ")
    code, _, _ = warning.partition(" ")
    result.setdefault(file_name, set()).add(code)

for file_name, errors in sorted(result.items()):
    print(f"    {file_name}: {', '.join(sorted(errors))}")
```

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -47,6 +47,7 @@ repos:
           - flake8-broken-line
           - flake8-comprehensions
           - pep8-naming
+          - flake8-docstrings
```
Comment (cclauss) [inline .pre-commit-config.yaml:50]: Please keep lists of dependencies in alphabetical order to make it easy to spot missing dependencies and difficult to add duplicates.
```suggestion
          - flake8-comprehensions
          - flake8-docstrings
          - pep8-naming
```

**Comment by cclauss** [comment]
Please fix git conflicts.

File: .flake8 [hunk 2]
```diff
@@ -6,3 +6,508 @@ extend-ignore =
     # Formatting style for `black`
     E203  # Whitespace before ':'
     W503  # Line break occurred before a binary operator
+    D104 # missing docstring in __ini__.py
```
Comment (CaedenPH) [inline .flake8:9]:
```suggestion
    D104  # Missing docstring in __init__.py
```

File: .flake8 [hunk 3]
```diff
@@ -6,3 +6,508 @@ extend-ignore =
...
+    D107 # Missing docstring in __init__
+    D301 # Use r""" if any backslashes in a docstring, some docstring couldn't be written without it
+    D2 # Shape of the docstring, like first line is summary and the next line is blank line
+    D4 # Style for the first line.  Sometimes we are lack of documentation and the first line is docstring.
```
Comment (CaedenPH) [inline .flake8:13]:
```suggestion
    D107  # Missing docstring in __init__
    D301  # Use r""" if docstring contains backslashes
    D2  # Shape of the docstring, like first line is summary and the next line is blank line
    D4  # Style for the first line.  Sometimes we are lack of documentation and the first line is docstring.
```

**Review by CaedenPH — CHANGES_REQUESTED** [review]
@cclauss I don't really like all the ignored files lacking docstrings, what is your opinion?
Perhaps after this pr is merged I will systematically implement docstring for these ignores.
I have a lot of exams coming up so it may take a while

**Comment by Cjkjvfnby** [comment]
Every other PR merged could make this one obsolete, because new undocumented things were added.

After merging this one I'll run extra checks to ensure that no unsuppressed warnings left.

**Comment by Cjkjvfnby** [comment]
Closing it since it became outdated.  It would be easy to recreate it from a scratch if needed.

**Pull request closed without merging**